component/attr_cache/attr_cache.go

```diff
@@ -276,29 +281,23 @@ func (ac *AttrCache) invalidateDirectory(path string) {
 	// Add a trailing / so that we only invalidate child paths under the directory and not paths that have the same prefix
 	prefix := dirToPrefix(path)
 
-	for key, value := range ac.cacheMap {
-		if strings.HasPrefix(key, prefix) {
-			// don't invalidate directories when cacheDirs is true
-			if ac.cacheDirs && value.attr.IsDir() {
-				continue
+	toBeInvalid, err := ac.cacheMap.get(prefix)
+	if err != nil {
+		log.Err("could not invalidate cached attr item due to the following error: ", err)
+	} else {
+		// don't invalidate directories when cacheDirs is true
+		if !ac.cacheDirs && toBeInvalid.attr.IsDir() {
```

Comment:
we don't want to invalidate folders that have cacheDirs as True though correct? if we invalidate with no conditions, we will invalidate the entire sub directory including folders that have cacheDirs set. Is that something we don't need to worry about?

Reply:
Yeah, I was wrong about the equivalency. Just calling invalidate would not give the same result.
I got excited, because the comments (which I wrote) right above the definition of `invalidateDirectory()` say this is not supposed to be called with the `ac.cacheDirs` flag high.
But it does get called with the flag high, in `SyncDir()`.
So yes, we still need this logic. It's just that the conditions are hard to read. ...which might be my fault for writing it this way.
I think the conditions can be simplified to this (and the extra nil check removed):
```
if !ac.cacheDirs {
	toBeInvalid.invalidate()
} else if !toBeInvalid.attr.IsDir() {
	toBeInvalid.invalidate()
} else {
	for _, childItem := range toBeInvalid.children {
		ac.invalidateDirectory(childItem.attr.Path)
	}
}
```